ocamllib/src/astlib.ml

```diff
@@ -0,0 +1,920 @@
+(* Include functions to be used by CInterface *)
+
+open Ast
+open Astutil
+open Type
+open Var
+
+exception SyntaxError of string
+exception IOError of string
+exception NotImplemented of unit
+exception InvalidAction of int
+
+type testType = int * int
+
+(*
+     Given an zippered AST, apply the action
+     Input:
+       - e : an AST with cursor (TODO: implement zast)
+       - action : action applied to e
+     Output:
+       the modified AST
+*)
+let change_ast (tree : Expr.z_t) (action : Action.t) : Expr.z_t =
+  (*handles actions on type subtrees:  
+      > uses three subfunctions to handle Move parent Move child and Construct
+      actions, respectively*)
+  let rec act_on_type (type_tree : Typ.z_t) : Typ.z_t =
+    
+    let build_type (subtr : Typ.t) (shape : Action.shape) : Typ.z_t =
+      (* actually creates the desired type at the cursor.*)
+      (*Also handles child wrapping logics *)
+      Cursor
+        (match shape with
+        | TypInt -> Typ.TInt
+        | TypBool -> Typ.TBool
+        | TypHole -> Typ.THole
+        | TypArrow_L -> Typ.TArrow (subtr, Typ.THole)
+        | TypArrow_R -> Typ.TArrow (Typ.THole, subtr)
+        | TypList -> Typ.TList subtr
+        | TypProd_L -> Typ.TProd (subtr, THole)
+        | TypProd_R -> Typ.TProd (THole, subtr)
+        | _ ->
+            subtr (* all other shapes are for exprssions which are not valid*))
```

Comment:
Zach wrote this part, so I'm not entirely familiar with this, but I believe that this is just to prevent the warning for non-exhaustive pattern matches.